packages/router/src/router.ts

```diff
@@ -672,7 +674,15 @@ export class Router {
 
                      filter(t => {
                        if (!t.guardsResult) {
-                         this.resetUrlToCurrentUrlTree();
+                         // The navigator change the location before triggered the popstate event,
+                         // so we need to go back to the current url if the guard result is false.
+                         if (t.source !== 'imperative') {
+                           const targetPagePosition =
+                               this.lastSuccessFulPageId - this.currentPageId;
```

Comment:
```suggestion
                               this.lastSuccessFulPageId - t.targetPageId;
```
I think this could be `t.tagetPageId` instead of `currentPageId`, right? That seems to make more sense gramatically: "Calculate the change in positions based on the difference between lastSuccessful and the target navigation"